neqo-transport/src/ecn.rs

```diff
@@ -0,0 +1,116 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use enum_map::EnumMap;
+use neqo_common::{qdebug, qinfo, qwarn, IpTosEcn};
+
+use crate::tracking::{EcnCount, PacketNumberSpace, SentPacket};
+
+/// The number of packets to use for testing a path for ECN capability.
+pub const ECN_TEST_COUNT: usize = 10;
+
+/// The state information related to testing a path for ECN capability.
+/// See RFC9000, Appendix A.4.
+#[derive(Debug, PartialEq, Default)]
+enum EcnValidationState {
+    /// The path is currently being tested for ECN capability.
+    #[default]
+    Testing,
+    /// The validation test has concluded but the path's ECN capability is not yet known.
+    Unknown,
+    /// The path is known to **not** be ECN capable.
+    Failed,
+    /// The path is known to be ECN capable.
+    Capable,
+}
+
+#[derive(Debug, Default)]
+pub struct EcnInfo {
+    /// The current state of ECN validation on this path.
+    state: EcnValidationState,
+    /// The number of packets sent so far on the path during the ECN validation.
+    sent: usize,
+    /// The ECN counts received in the last ACK on this path, for each packet number space.
+    /// Won't be updated after ECN has been tested on the path.
+    count: EnumMap<PacketNumberSpace, EcnCount>,
+}
+
+impl EcnInfo {
+    /// Count the number of packets sent out on this path during ECN validation.
+    /// Exit ECN validation if the number of packets sent exceeds `ECN_TEST_COUNT`.
+    /// We do not implement the part of the RFC that says to exit ECN validation if the time since
+    /// the start of ECN validation exceeds 3 * PTO, since this seems to happen much too quickly.
+    pub fn count_packets_out(&mut self) {
+        if self.state == EcnValidationState::Testing {
+            if self.sent < ECN_TEST_COUNT {
+                self.sent += 1;
+            } else {
+                qdebug!("ECN probing concluded with {} packet sent", self.sent);
+                self.state = EcnValidationState::Unknown;
+            }
+        }
+    }
+
+    /// After the ECN validation test has ended, check if the path is ECN capable.
+    pub fn validate_ack_ecn(
+        &mut self,
+        space: PacketNumberSpace,
+        acked_packets: &[SentPacket],
+        ecn_count: &EcnCount,
+    ) {
+        // RFC 9000, Appendix A.4:
+        // From the "unknown" state, successful validation of the ECN counts in an ACK frame
+        // (see Section 13.4.2.1) causes the ECN state for the path to become "capable", unless
+        // no marked packet has been acknowledged.
+        if self.state != EcnValidationState::Unknown {
+            return;
+        }
```

Comment:
The problem with ECN validation is that it is hard to reset the state.  So performing this validation constantly is not going to work with the code in its current state.  The reason for that is that a single bleached marking will always manifest in the counts.

That's why this doesn't re-test or continuously check for success: this will just fail.

A more robust approach would involve re-testing a failed path, but you have to deliberately switch back to a testing state, plus snapshot the counts at that time, and account for any packets that are still in flight.  It's a bit tricky.  Better to just accept that ECN doesn't work.

Reply:
I think there is a misunderstanding here @martinthomson.

I am suggesting to support going from `EcnValidationState::Capable` to `EcnValidationState::Failed`. In other words, once in `Capable` I am suggesting to still continuously check ECN validity, potentially going to `Failed` on failure.  Once in `EcnValidationState::Failed`, there is no going back, i.e. no re-testing.

If I understand your comment above, you are saying that re-testing, i.e. going from `EcnValidationState::Failed` eventually to `EcnValidationState::Capable` is not going to work with the current logic, but instead requires a more involved approach. I agree with that.